cpp/src/arrow/sparse_tensor_test.cc

```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+    return std::make_shared<SparseCSFTensor>(si,
+                                             CTypeTraits<CValueType>::type_singleton(),
+                                             data, this->shape_, this->dim_names_);
+  }
+};
+
+TYPED_TEST_CASE_P(TestSparseCSFTensorForIndexValueType);
+
+TYPED_TEST_P(TestSparseCSFTensorForIndexValueType, TestBufferToSparseTensor) {
+  using IndexValueType = TypeParam;
+  using c_index_value_type = typename IndexValueType::c_type;
+
+  std::vector<std::vector<c_index_value_type>> indptr_values = {
+      {0, 2, 3}, {0, 1, 3, 4}, {0, 2, 4, 5, 8}};
+  std::vector<std::vector<c_index_value_type>> indices_values = {
+      {1, 2}, {1, 2, 2}, {1, 1, 2, 2}, {2, 3, 1, 3, 1, 1, 2, 3}};
+  std::vector<int64_t> indices_shapes = {2, 3, 4, 8};
+  std::vector<int64_t> axis_order = {0, 1, 2, 3};
+  std::vector<int64_t> sparse_values = {1, 2, 3, 4, 5, 6, 7, 8};
+  std::vector<int64_t> shape = {3, 3, 3, 4};
+  std::vector<std::string> dim_names = {"a", "b", "c", "d"};
+
+  std::shared_ptr<Buffer> data_buffer = Buffer::Wrap(sparse_values);
+  std::vector<std::shared_ptr<Buffer>> indptr_buffers;
+  std::vector<std::shared_ptr<Buffer>> indices_buffers;
+  for (auto& indptr : indptr_values) indptr_buffers.push_back(Buffer::Wrap(indptr));
+  for (auto& indices : indices_values) indices_buffers.push_back(Buffer::Wrap(indices));
+
+  std::shared_ptr<SparseCSFIndex> si;
+  ASSERT_OK_AND_ASSIGN(
+      si,
+      SparseCSFIndex::Make(TypeTraits<IndexValueType>::type_singleton(), indices_shapes,
+                           axis_order, indptr_buffers, indices_buffers));
+  std::shared_ptr<SparseCSFTensor> st =
+      std::make_shared<SparseCSFTensor>(si, int64(), data_buffer, shape, dim_names);
+
+  ASSERT_TRUE(st->Equals(*this->sparse_tensor_from_dense_));
+}
+
+TYPED_TEST_P(TestSparseCSFTensorForIndexValueType, TestTensorToSparseTensor) {
+  using IndexValueType = TypeParam;
+  std::vector<int64_t> shape = {3, 3, 3, 4};
+  std::vector<int64_t> values = {
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 4, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 7, 8};
+
+  std::shared_ptr<Buffer> buffer = Buffer::Wrap(values);
+  Tensor tensor(int64(), buffer, shape, {}, this->dim_names_);
+
+  std::shared_ptr<SparseCSFTensor> sparse_tensor;
+  ASSERT_OK_AND_ASSIGN(
+      sparse_tensor,
+      SparseCSFTensor::Make(tensor, TypeTraits<IndexValueType>::type_singleton()));
+
+  ASSERT_EQ(8, sparse_tensor->non_zero_length());
+  ASSERT_TRUE(sparse_tensor->is_mutable());
+  ASSERT_TRUE(sparse_tensor->Equals(*this->sparse_tensor_from_dense_));
+}
+
+TYPED_TEST_P(TestSparseCSFTensorForIndexValueType, TestSparseTensorToTensor) {
+  std::vector<int64_t> shape = {3, 3, 3, 4};
```

Comment:
Can you add a test with a non-ascending shape?

Reply:
Added `TestNonAscendingShape`.